File: backend/app/schemas/workflow.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel

from app.models.workflow import WorkflowStatus


class WorkflowCreate(BaseModel):
    name: str
    description: str = ""
    definition: dict[str, Any] = {"nodes": [], "edges": []}
    status: WorkflowStatus = WorkflowStatus.DRAFT


class WorkflowUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    definition: dict[str, Any] | None = None
    status: WorkflowStatus | None = None


class WorkflowRead(BaseModel):
    id: str
    name: str
    description: str
    definition: dict[str, Any]
    version: int
    status: WorkflowStatus
    created_at: datetime
    updated_at: datetime


class ValidationIssue(BaseModel):
    code: str
    message: str
    node_id: str | None = None
    edge: dict[str, Any] | None = None


class ValidationResult(BaseModel):
    valid: bool
    errors: list[ValidationIssue] = []
    warnings: list[ValidationIssue] = []
    topo_order: list[str] = []


class ExecuteRequest(BaseModel):
    input_payload: dict[str, Any] = {}

File: frontend/src/types/workflow.ts
export type WorkflowStatus = 'draft' | 'published'
export type ConditionOp = 'eq' | 'neq' | 'gt' | 'gte' | 'lt' | 'lte' | 'in' | 'contains'

export interface WorkflowEdgeCondition {
  field: string
  op: ConditionOp
  value: unknown
}

export interface WorkflowNode {
  id: string
  agent_id: string
  label?: string
  input_mapping?: Record<string, string>
}

export interface WorkflowEdge {
  from: string
  to: string
  condition?: WorkflowEdgeCondition | null
}

export interface WorkflowDefinition {
  nodes: WorkflowNode[]
  edges: WorkflowEdge[]
}

export interface Workflow {
  id: string
  name: string
  description: string
  definition: WorkflowDefinition
  version: number
  status: WorkflowStatus
  created_at: string
  updated_at: string
}

export interface WorkflowCreate {
  name: string
  description?: string
  definition: WorkflowDefinition
  status?: WorkflowStatus
}

export type WorkflowUpdate = Partial<WorkflowCreate>

export interface ValidationIssue {
  code: string
  message: string
  node_id?: string | null
  edge?: Record<string, unknown> | null
}

export interface ValidationResult {
  valid: boolean
  errors: ValidationIssue[]
  warnings: ValidationIssue[]
  topo_order: string[]
}
